### ISSUE TITLE
[Feature Request] Nested sub-menus

### ISSUE THREAD

**Issue by umar-stride-studio** [notDmDrl/pull_down_button#26]
I want to implement something like this in  flutter application

![IMG_1419](https://github.com/notDmDrl/pull_down_button/assets/116542608/2951f5df-1430-4751-8fb8-86995ec60002)

![IMG_1420](https://github.com/notDmDrl/pull_down_button/assets/116542608/bb527619-fd10-4e7f-a113-c215de5e5df3)

You can see when user click on sort by button another menu appeared on previous menu. I want to implement same functionality. Let me know if it is possible. If not then any alternative for it?

**Comment by umar-stride-studio** [comment]
@iSa1vatore @notDmDrl

**Comment by notDmDrl** [comment]
Hi @umar-stride-studio 

This is yet to be implemented, so no, this is not possible. 

I don't have any estimates on when this will become available since, as seen in https://github.com/flutter/flutter/issues/120960, in https://github.com/flutter/flutter/issues/60298#issuecomment-1483754218 and https://github.com/flutter/flutter/issues/60298#issuecomment-1513586626, @davidhicks980 is working on it already and plans to submit a PR (whether it will be to this repo, Flutter or his package).

I want to wait on said PR so there is no redundant work done by me 😃

If PR ends up here, it will be released ASAP. If not, I'll begin to work on my implementation.

As for an alternative, either separate the submenu to its own `PullDownButton` or inline it in an existing one.

---

Sorry for the ping @davidhicks980 :slightly_smiling_face:

**Comment by davidhicks980** [comment]
No worries! I'm actually on vacation and working on this at the moment. To accomodate the expansion panel, I had to change the API surface a bit, so I will attempt a draft PR in the Flutter repo, but, if some of the more sophisticated features like expansion get rejected in the main repo (I imagine some will because of the testing complexity), I was planning on submitting pull requests here for anyone who wants more feature parity. Also -- off topic -- thank you for the continuous swipe idea! I added it to the nested menus.

 I _think_ I will have a draft PR in the main repo by Monday, supposing no road blocks come up. Flutter's popup menus have quite a few tests, so we will see. I'll be sure to tag you once it is live.


https://github.com/notDmDrl/pull_down_button/assets/59215665/44af3dd6-819f-49f0-9ed9-a5334e0415fe


https://github.com/notDmDrl/pull_down_button/assets/59215665/2f3c93d0-bd4f-4fad-ad4c-6ec9b4e2da41

**Comment by umar-stride-studio** [comment]
@davidhicks980 when i will get an update regarding this feature

**Comment by umar-stride-studio** [comment]
@notDmDrl @davidhicks980 any update?

**Comment by davidhicks980** [comment]
https://github.com/davidhicks980/cupertino_menu I had to remove the nested menu navigation temporarily and there are some styling issues, but I think the example should provide some guidance on how to implement.

**Comment by notDmDrl** [comment]
Hi @alterhuman

Both solutions (mine and David's) are pretty different in structure and internal API design, so it will be a rather non-trivial task to merge his solution to this repo.

Unfortunately, due to some IRL stuff, I will not be able to work on my implementation for a few months so if anyone wants to give this issue a try and submit a PR it would be nice 😃

**Comment by davidhicks980** [comment]
@notDmDrl  @alterhuman Same with IRL stuff, but I'm actually about to have a lot less to do, so I may be able to rework it into this repo. Nested menus pose a lot of challenges, particularly with regard to things like hovering sliding, communicating between layers, and responding to layout changes between layers. If I'm remembering correctly, I think the repo I shared used a nested navigator approach, which is probably not the best way to do it. You can use nested overlays, but tracking changes between layers is difficult (composited transform follower has it's own tradeoffs). As far as I can tell, Flutter doesn't provides a great api for working with nested layouts because layers really shouldn't affect each other (i.e. single pass), so I've been meaning to shift to a multi child layout. As this is meant to be full-match Flutter's way of doing things (@notDmDrl gives a lot more customization options than the Cupertino theme does), the repo I shared doesn't neatly slot in to this repo. That being said, once I have a solid approach, an actual migration shouldn't be too tough. 

Long story short -- on the way, but I've been busy with work and life.

**Comment by manudicri** [comment]
Hi, any news on this? That could be really awesome

**Comment by davidhicks980** [comment]
Dang it's been a year. This turned into a bigger project to change some of flutters internal menus, including producing a raw menu anchor for the flutter repo. So it's been slow but moving along. Some of the changes have already landed, but the Cupertino menu is still in the queue.

**Comment by davidhicks980** [comment]
@namanshergill Mostly, yes. I'll start integrating the changes.

**Comment by notDmDrl** [comment]
Congrats on the progress @davidhicks980 😃

> Also, @notDmDrl, thanks again for inspiring this project. It took a bit longer than expected.

Maaan, I don't even want to get a single ounce of credit for inspiring you 😅 Your work just blows my package out of the water :). Eager to see it all being merged.

> Does that mean this feature could be implemented now?

Well, yes, but also no. The package deliberately follows a minimum Flutter version to be similar to most Flutter's packages (which are usually a few stable releases back from the latest). So the recently merged PR will not be available for us for quite some time. 

However, I think that there is a chance that I'll just bump the minimum Flutter version to the latest stable and use all the latest features available. But this will probably happen when I deprecate the package after @davidhicks980's work on menus is over.

**Comment by oz888z** [comment]
N
